// nivo/lib/components/scales/XYScales.js
/*
 * This file is part of the nivo project.
 *
 * (c) Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _d = require('d3');

var _d2 = _interopRequireDefault(_d);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var XYScales = function (_Component) {
    _inherits(XYScales, _Component);

    function XYScales() {
        _classCallCheck(this, XYScales);

        return _possibleConstructorReturn(this, (XYScales.__proto__ || Object.getPrototypeOf(XYScales)).apply(this, arguments));
    }

    _createClass(XYScales, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                data = _props.data,
                identityAccessor = _props.identityAccessor,
                valueAccessor = _props.valueAccessor,
                width = _props.width,
                height = _props.height,
                children = _props.children;


            if (_react2.default.Children.count(children) === 0) {
                return null;
            }

            var xScale = _d2.default.scale.linear().range([0, width]).domain([0, data.length - 1]);

            var yScale = _d2.default.scale.linear().range([height, 0]).domain([0, _d2.default.max(data.map(valueAccessor))]);

            return _react2.default.createElement(
                'g',
                null,
                _react2.default.Children.map(children, function (child) {
                    return _react2.default.cloneElement(child, { data: data.slice(), xScale: xScale, yScale: yScale, width: width, height: height });
                })
            );
        }
    }]);

    return XYScales;
}(_react.Component);

XYScales.displayName = 'XYScales';

var array = _react.PropTypes.array,
    func = _react.PropTypes.func,
    number = _react.PropTypes.number;


XYScales.propTypes = {
    data: array.isRequired,
    identityAccessor: func.isRequired,
    valueAccessor: func.isRequired,
    width: number,
    height: number
};

XYScales.defaultProps = {
    identityAccessor: function identityAccessor(d, i) {
        return i;
    },
    valueAccessor: function valueAccessor(d) {
        return d;
    }
};

exports.default = XYScales;
