// nivo/lib/components/charts/Markers.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _ColorUtils = require('../../../ColorUtils');

var _propertiesConverters = require('../../../lib/propertiesConverters');

var _BarItem = require('./BarItem');

var _BarItem2 = _interopRequireDefault(_BarItem);

var _BarItemLabel = require('./BarItemLabel');

var _BarItemLabel2 = _interopRequireDefault(_BarItemLabel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Markers = function (_Component) {
    _inherits(Markers, _Component);

    function Markers() {
        _classCallCheck(this, Markers);

        return _possibleConstructorReturn(this, (Markers.__proto__ || Object.getPrototypeOf(Markers)).apply(this, arguments));
    }

    _createClass(Markers, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                data = _props.data,
                scales = _props.scales,
                _xScale = _props.xScale,
                _yScale = _props.yScale,
                width = _props.width,
                height = _props.height,
                color = _props.color,
                x = _props.x,
                y = _props.y;


            var xScale = scales[_xScale];
            var yScale = scales[_yScale];

            var getX = _lodash2.default.isFunction(x) ? x : function (d) {
                return d[x];
            };
            var getY = _lodash2.default.isFunction(y) ? y : function (d) {
                return d[y];
            };

            var rects = [];
            data.forEach(function (d, i) {
                var x = void 0;
                var y = void 0;
                var barWidth = void 0;
                var barHeight = void 0;

                if (xScale.bandwidth) {
                    barWidth = xScale.bandwidth(); // / series.length
                    x = xScale(getX(d)); // + barWidth * serieIndex
                } else {
                    x = 0;
                    barWidth = xScale(getX(d));
                }

                if (yScale.bandwidth) {
                    barHeight = yScale.bandwidth(); // / series.length
                    y = yScale(getY(d)); // + barHeight * serieIndex
                } else {
                    y = yScale(getY(d));
                    barHeight = height - y;
                }

                if (barWidth > 0 && barHeight > 0) {
                    rects.push({
                        key: 'bar.' + i,
                        //value,
                        x: x, y: y,
                        width: barWidth,
                        height: barHeight,
                        color: color });
                }
            });

            console.log(rects);

            return _react2.default.createElement(
                'g',
                { className: 'bars' },
                rects.map(function (d) {
                    return _react2.default.createElement(_BarItem2.default, _extends({}, d, { key: d.key }));
                })
            );
            /*
            const {
                data,
                groupMode,
                identity: _identity,
                margin: _margin,
                width: _width, height: _height,
                colors,
                scales: _scales,
                series,
                axes,
                enableGridX, enableGridY,
                enableLabels,
                animate, motionStiffness, motionDamping,
            } = this.props
             const margin = Object.assign({}, Nivo.defaults.margin, _margin)
            const width  = _width  - margin.left - margin.right
            const height = _height - margin.top - margin.bottom
             const color  = getColorRange(colors)
             const rects = []
            series.forEach((serie, serieIndex) => {
                const xScale = scales[serie.xScale]
                const yScale = scales[serie.yScale]
                 const getX = _.isFunction(serie.x) ? serie.x : d => d[serie.x]
                const getY = _.isFunction(serie.y) ? serie.y : d => d[serie.y]
                 data.forEach((d, i) => {
                    let x
                    let y
                    let barWidth
                    let barHeight
                     if (xScale.bandwidth) {
                        barWidth = xScale.bandwidth() / series.length
                        x        = xScale(getX(d)) + barWidth * serieIndex
                    } else {
                        x        = 0
                        barWidth = xScale(getX(d))
                    }
                     if (yScale.bandwidth) {
                        barHeight = yScale.bandwidth() / series.length
                        y         = yScale(getY(d)) + barHeight * serieIndex
                    } else {
                        y         = yScale(getY(d))
                        barHeight = height - y
                    }
                     if (barWidth > 0 && barHeight > 0) {
                        rects.push({
                            key:    `${serieIndex}.${i}`,
                            //value,
                            x, y,
                            width:  barWidth,
                            height: barHeight,
                            color:  color(serieIndex),
                        })
                    }
                })
            })
             /*
            if (groupMode === 'stacked') {
                const series = stack().keys(keys)(data)
                 y.domain([0, max(series[series.length - 1], d => d[1])])
                 series.forEach(serie => {
                    serie.forEach(d => {
                        rects.push({
                            key:    `${serie.key}.${identity(d.data)}`,
                            value:  d[1],
                            x:      x(identity(d.data)),
                            y:      y(d[1]),
                            width:  x.bandwidth(),
                            height: y(d[0]) - y(d[1]),
                            color:  color(serie.key),
                        })
                    })
                })
            } else {
                y.domain([0, max(data, d => max(_.values(_.pick(d, keys))))])
                 const barWidth = x.bandwidth() / keys.length
                 data.forEach(bucket => {
                    let index = 0
                    _.forOwn(_.pick(bucket, keys), (value, key) => {
                        rects.push({
                            key:    `${key}.${identity(bucket)}`,
                            value,
                            x:      x(identity(bucket)) + barWidth * index,
                            y:      y(value),
                            width:  barWidth,
                            height: height - y(value),
                            color:  color(key),
                        })
                        index++
                    })
                })
            }
             const { props } = this
             return (
                <svg xmlns="http://www.w3.org/2000/svg" className="nivo_bar" width={_width} height={_height}>
                    <g
                        className="nivo_bar_wrapper"
                        transform={`translate(${margin.left},${margin.top})`}
                    >
                        {rects.map(d => {
                            return (
                                <BarItem {...d} key={d.key} />
                            )
                        })}
                        {enableLabels && rects.map(d => {
                            return (
                                <BarItemLabel {...d} key={d.key} />
                            )
                        })}
                        {['top', 'right', 'bottom', 'left'].map(orient => {
                            return getAxis(axes, scales, orient, width, height, props)
                        })}
                    </g>
                </svg>
            )
             */
        }
    }]);

    return Markers;
}(_react.Component);

Markers.propTypes = {
    colors: _react.PropTypes.any.isRequired,
    scales: _react.PropTypes.object,
    width: _react.PropTypes.number,
    height: _react.PropTypes.number,
    // labels
    enableLabels: _react.PropTypes.bool.isRequired,
    // motion
    animate: _react.PropTypes.bool.isRequired,
    motionStiffness: _react.PropTypes.number.isRequired,
    motionDamping: _react.PropTypes.number.isRequired
};
Markers.defaultProps = {
    colors: _Nivo2.default.defaults.colorRange,
    // labels
    enableLabels: true,
    // motion
    animate: true,
    motionStiffness: _Nivo2.default.defaults.motionStiffness,
    motionDamping: _Nivo2.default.defaults.motionDamping
};
exports.default = Markers;
