// mozaik-ext-github/lib/components/pull-requests/PullRequest.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _ui = require('mozaik/ui');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PullRequest = function (_Component) {
    _inherits(PullRequest, _Component);

    function PullRequest() {
        _classCallCheck(this, PullRequest);

        return _possibleConstructorReturn(this, (PullRequest.__proto__ || Object.getPrototypeOf(PullRequest)).apply(this, arguments));
    }

    _createClass(PullRequest, [{
        key: 'render',
        value: function render() {
            var pullRequest = this.props.pullRequest;
            var title = pullRequest.title,
                html_url = pullRequest.html_url,
                created_at = pullRequest.created_at,
                user = pullRequest.user;


            return _react2.default.createElement(_ui.WidgetListItem, {
                title: _react2.default.createElement(
                    'span',
                    null,
                    _react2.default.createElement(
                        'a',
                        { href: html_url, target: '_blank' },
                        title
                    ),
                    ' by ',
                    _react2.default.createElement(
                        'a',
                        { href: user.html_url, target: '_blank' },
                        user.login
                    )
                ),
                pre: _react2.default.createElement(
                    _ui.WidgetAvatar,
                    { href: user.html_url, size: '4vmin' },
                    _react2.default.createElement('img', { src: user.avatar_url, alt: user.login })
                ),
                meta: _react2.default.createElement(
                    'span',
                    null,
                    _react2.default.createElement('i', { className: 'fa fa-clock-o' }),
                    '\xA0',
                    (0, _moment2.default)(created_at).fromNow()
                )
            });
        }
    }]);

    return PullRequest;
}(_react.Component);

PullRequest.propTypes = {
    pullRequest: _react.PropTypes.shape({
        title: _react.PropTypes.string.isRequired,
        html_url: _react.PropTypes.string.isRequired,
        created_at: _react.PropTypes.string.isRequired,
        user: _react.PropTypes.shape({
            html_url: _react.PropTypes.string.isRequired,
            avatar_url: _react.PropTypes.string.isRequired,
            login: _react.PropTypes.string.isRequired
        }).isRequired
    }).isRequired
};
exports.default = PullRequest;
