// mozaik-ext-github/lib/components/pull-requests/PullRequests.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _PullRequest = require('./PullRequest');

var _PullRequest2 = _interopRequireDefault(_PullRequest);

var _ui = require('mozaik/ui');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PullRequests = function (_Component) {
    _inherits(PullRequests, _Component);

    function PullRequests() {
        _classCallCheck(this, PullRequests);

        return _possibleConstructorReturn(this, (PullRequests.__proto__ || Object.getPrototypeOf(PullRequests)).apply(this, arguments));
    }

    _createClass(PullRequests, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                repository = _props.repository,
                title = _props.title,
                apiData = _props.apiData,
                apiError = _props.apiError;


            var body = _react2.default.createElement(_ui.WidgetLoader, null);
            var count = void 0;
            if (apiData) {
                count = apiData.pullRequests.length;
                body = _react2.default.createElement(
                    'div',
                    null,
                    apiData.pullRequests.map(function (pullRequest) {
                        return _react2.default.createElement(_PullRequest2.default, { key: pullRequest.id, pullRequest: pullRequest });
                    })
                );
            }

            return _react2.default.createElement(
                _ui.Widget,
                null,
                _react2.default.createElement(_ui.WidgetHeader, {
                    title: title || 'Pull Requests',
                    subject: title ? null : repository,
                    count: count,
                    icon: null
                }),
                _react2.default.createElement(
                    _ui.WidgetBody,
                    null,
                    _react2.default.createElement(
                        _ui.TrapApiError,
                        { error: apiError },
                        body
                    )
                )
            );
        }
    }], [{
        key: 'getApiRequest',
        value: function getApiRequest(_ref) {
            var repository = _ref.repository;

            return {
                id: 'github.pullRequests.' + repository,
                params: { repository: repository }
            };
        }
    }]);

    return PullRequests;
}(_react.Component);

PullRequests.propTypes = {
    repository: _react.PropTypes.string.isRequired,
    title: _react.PropTypes.string,
    apiData: _react.PropTypes.shape({
        pullRequests: _react.PropTypes.arrayOf(_react.PropTypes.object).isRequired
    }),
    apiError: _react.PropTypes.object
};
exports.default = PullRequests;
