// nivo/lib/components/charts/line/Line.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _curve = require('../../../properties/curve');

var _d = require('d3');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Line = function (_Component) {
    _inherits(Line, _Component);

    function Line() {
        _classCallCheck(this, Line);

        return _possibleConstructorReturn(this, (Line.__proto__ || Object.getPrototypeOf(Line)).apply(this, arguments));
    }

    _createClass(Line, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                data = _props.data,
                scales = _props.scales,
                _xScale = _props.xScale,
                _yScale = _props.yScale,
                x = _props.x,
                y = _props.y,
                curve = _props.curve;


            var xScale = scales[_xScale];
            var yScale = scales[_yScale];

            var getX = _lodash2.default.isFunction(x) ? x : function (d) {
                return d[x];
            };
            var getY = _lodash2.default.isFunction(y) ? y : function (d) {
                return d[y];
            };

            var points = data.map(function (d) {
                return {
                    x: xScale(getX(d)),
                    y: yScale(getY(d))
                };
            });

            var lineGenerator = (0, _d.line)().x(function (d) {
                return d.x;
            }).y(function (d) {
                return d.y;
            }).curve(_curve.curvePropMapping[curve]);

            return _react2.default.createElement('path', {
                d: lineGenerator(points),
                fill: 'none',
                strokeWidth: 2,
                stroke: '#fff'
            });
        }
    }]);

    return Line;
}(_react.Component);

Line.propTypes = {
    scales: _react.PropTypes.object.isRequired,
    curve: _curve.curvePropType.isRequired
};
Line.defaultProps = {
    scales: {},
    curve: 'linear'
};
exports.default = Line;
