// nivo/lib/components/charts/radial-stack/RadialStack.js
/*
 * This file is part of the nivo project.
 *
 * (c) Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _RadialStackD3Svg = require('../../../lib/charts/radial-stack/RadialStackD3Svg');

var _RadialStackD3Svg2 = _interopRequireDefault(_RadialStackD3Svg);

var _decoratorsFromReactChildren = require('../../../lib/decoratorsFromReactChildren');

var _decoratorsFromReactChildren2 = _interopRequireDefault(_decoratorsFromReactChildren);

var _PropTypes = require('../../../PropTypes');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RADIAL_STACK_DECORATOR = 'decorateRadialStack';

var RadialStack = function (_Component) {
    _inherits(RadialStack, _Component);

    function RadialStack() {
        _classCallCheck(this, RadialStack);

        return _possibleConstructorReturn(this, (RadialStack.__proto__ || Object.getPrototypeOf(RadialStack)).apply(this, arguments));
    }

    _createClass(RadialStack, [{
        key: 'shouldComponentUpdate',
        value: function shouldComponentUpdate(nextProps) {
            this.radialStack.decorate((0, _decoratorsFromReactChildren2.default)(nextProps.children, RADIAL_STACK_DECORATOR));
            this.radialStack.draw(_lodash2.default.assign({}, nextProps, {
                margin: _lodash2.default.assign({}, _Nivo2.default.defaults.margin, nextProps.margin)
            }));

            return false;
        }
    }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.radialStack = (0, _RadialStackD3Svg2.default)((0, _reactDom.findDOMNode)(this));
            this.radialStack.decorate((0, _decoratorsFromReactChildren2.default)(this.props.children, RADIAL_STACK_DECORATOR));
            this.radialStack.draw(_lodash2.default.assign({}, this.props, {
                margin: _lodash2.default.assign({}, _Nivo2.default.defaults.margin, this.props.margin)
            }));
        }
    }, {
        key: 'render',
        value: function render() {
            return _react2.default.createElement('svg', { className: 'nivo_radial-stack' });
        }
    }]);

    return RadialStack;
}(_react.Component);

var array = _react.PropTypes.array,
    number = _react.PropTypes.number,
    string = _react.PropTypes.string,
    func = _react.PropTypes.func,
    any = _react.PropTypes.any,
    oneOf = _react.PropTypes.oneOf;


RadialStack.propTypes = {
    width: number.isRequired,
    height: number.isRequired,
    margin: _PropTypes.margin,
    innerRadius: number.isRequired,
    sort: func,
    layers: array.isRequired,
    offset: oneOf(['silhouette', 'wiggle', 'expand', 'zero']).isRequired,
    keyProp: string.isRequired,
    valueProp: string.isRequired,
    colors: any.isRequired,
    transitionDuration: number.isRequired,
    transitionEasing: string.isRequired
};

RadialStack.defaultProps = {
    margin: _Nivo2.default.defaults.margin,
    innerRadius: .6,
    sort: null,
    offset: 'zero',
    keyProp: 'label',
    valueProp: 'value',
    transitionDuration: _Nivo2.default.defaults.transitionDuration,
    transitionEasing: _Nivo2.default.defaults.transitionEasing,
    colors: _Nivo2.default.defaults.colorRange
};

exports.default = RadialStack;
