// nivo/lib/components/charts/calendar/CalendarProps.js
/*
 * This file is part of the nivo project.
 *
 * (c) Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.calendarDefaultProps = exports.calendarPropTypes = undefined;

var _react = require('react');

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _PropTypes = require('../../../PropTypes');

var _directions = require('../../../constants/directions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var number = _react.PropTypes.number,
    string = _react.PropTypes.string,
    bool = _react.PropTypes.bool,
    func = _react.PropTypes.func,
    shape = _react.PropTypes.shape,
    oneOf = _react.PropTypes.oneOf,
    oneOfType = _react.PropTypes.oneOfType,
    instanceOf = _react.PropTypes.instanceOf,
    arrayOf = _react.PropTypes.arrayOf;

/**
 * Calendar components propTypes.
 *
 * @type {object}
 */

var calendarPropTypes = exports.calendarPropTypes = {
    width: number.isRequired,
    height: number.isRequired,
    margin: _PropTypes.margin,
    from: oneOfType([string, instanceOf(Date)]).isRequired,
    to: oneOfType([string, instanceOf(Date)]).isRequired,
    data: arrayOf(shape({
        day: string.isRequired,
        value: number.isRequired
    })).isRequired,
    onDayClick: func.isRequired,
    direction: oneOf([_directions.DIRECTION_HORIZONTAL, _directions.DIRECTION_VERTICAL]),
    colorScale: _PropTypes.scale.isRequired,
    emptyColor: string.isRequired,
    // years
    yearSpacing: number.isRequired,
    yearLegendOffset: number.isRequired,
    // days
    daySpacing: number.isRequired,
    dayBorderWidth: number.isRequired,
    dayBorderColor: string.isRequired,
    // months
    monthBorderWidth: number.isRequired,
    monthBorderColor: string.isRequired,
    monthLegendOffset: number.isRequired,
    // transitions
    animate: bool.isRequired,
    motionStiffness: number.isRequired, // react-motion
    motionDamping: number.isRequired, // react-motion
    transitionDuration: number.isRequired, // d3 transitions
    transitionEasing: string.isRequired, // d3 transitions
    transitionStaggering: number.isRequired };

/**
 * Calendar components defaultProps.
 *
 * @type {object}
 */
var calendarDefaultProps = exports.calendarDefaultProps = {
    margin: _Nivo2.default.defaults.margin,
    direction: _directions.DIRECTION_HORIZONTAL,
    onDayClick: function onDayClick() {},
    emptyColor: '#fff',
    // years
    yearSpacing: 30,
    yearLegendOffset: 10,
    // days
    daySpacing: 0,
    dayBorderWidth: 1,
    dayBorderColor: '#000',
    // months
    monthBorderWidth: 2,
    monthBorderColor: '#000',
    monthLegendOffset: 6,
    // transitions
    animate: false,
    motionStiffness: _Nivo2.default.defaults.motionStiffness, // react-motion
    motionDamping: _Nivo2.default.defaults.motionDamping, // react-motion
    transitionDuration: _Nivo2.default.defaults.transitionDuration, // d3 transitions
    transitionEasing: _Nivo2.default.defaults.transitionEasing, // d3 transitions
    transitionStaggering: 5 };
