// nivo/lib/components/charts/bubble/BubbleD3.js
/*
 * This file is part of the nivo project.
 *
 * (c) Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _BubbleProps = require('./BubbleProps');

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _BubbleD3Svg = require('../../../lib/charts/bubble/BubbleD3Svg');

var _BubbleD3Svg2 = _interopRequireDefault(_BubbleD3Svg);

var _decoratorsFromReactChildren = require('../../../lib/decoratorsFromReactChildren');

var _decoratorsFromReactChildren2 = _interopRequireDefault(_decoratorsFromReactChildren);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BubbleD3 = function (_Component) {
    _inherits(BubbleD3, _Component);

    function BubbleD3() {
        _classCallCheck(this, BubbleD3);

        return _possibleConstructorReturn(this, (BubbleD3.__proto__ || Object.getPrototypeOf(BubbleD3)).apply(this, arguments));
    }

    _createClass(BubbleD3, [{
        key: 'renderChart',
        value: function renderChart(props) {
            this.bubble.decorate((0, _decoratorsFromReactChildren2.default)(props.children, 'decorateBubble'));
            this.bubble.draw(_lodash2.default.assign({}, props, {
                margin: _lodash2.default.assign({}, _Nivo2.default.defaults.margin, props.margin)
            }));
        }
    }, {
        key: 'shouldComponentUpdate',
        value: function shouldComponentUpdate(nextProps) {
            this.renderChart(nextProps);

            return false;
        }
    }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.bubble = (0, _BubbleD3Svg2.default)((0, _reactDom.findDOMNode)(this.refs.svg));
            this.renderChart(this.props);
        }
    }, {
        key: 'render',
        value: function render() {
            return _react2.default.createElement('svg', { ref: 'svg', className: 'nivo_bubble' });
        }
    }]);

    return BubbleD3;
}(_react.Component);

BubbleD3.propTypes = _lodash2.default.omit(_BubbleProps.bubblePropTypes, ['children', 'namespace', 'motionStiffness', 'motionDamping']);

BubbleD3.defaultProps = _lodash2.default.omit(_BubbleProps.bubbleDefaultProps, ['namespace', 'motionStiffness', 'motionDamping']);

exports.default = BubbleD3;
