// nivo/lib/components/charts/bubble/BubbleProps.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bubbleDefaultProps = exports.bubblePropTypes = undefined;

var _react = require('react');

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _PropTypes = require('../../../PropTypes');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    object = _react.PropTypes.object,
    number = _react.PropTypes.number,
    string = _react.PropTypes.string,
    any = _react.PropTypes.any,
    func = _react.PropTypes.func,
    oneOf = _react.PropTypes.oneOf;

/**
 * Bubble components propTypes.
 *
 * @type {object}
 */

var bubblePropTypes = exports.bubblePropTypes = {
  width: number.isRequired,
  height: number.isRequired,
  margin: _PropTypes.margin,
  root: object.isRequired,
  leavesOnly: bool.isRequired,
  onBubbleClick: func.isRequired,
  identity: string.isRequired,
  value: string.isRequired,
  padding: number.isRequired,
  colors: any.isRequired,
  // placeholders
  namespace: oneOf(['html', 'svg']),
  children: func.isRequired,
  // border
  borderWidth: number.isRequired,
  borderColor: any.isRequired,
  // labels
  enableLabel: bool.isRequired,
  label: string.isRequired,
  labelFormat: string,
  labelTextColor: any.isRequired,
  labelTextDY: number.isRequired,
  labelSkipRadius: number.isRequired,
  // transitions
  animate: bool.isRequired,
  motionStiffness: number.isRequired, // react-motion
  motionDamping: number.isRequired, // react-motion
  transitionDuration: number.isRequired, // d3 transitions
  transitionEasing: string.isRequired };

/**
 * Bubble components defaultProps.
 *
 * @type {object}
 */
var bubbleDefaultProps = exports.bubbleDefaultProps = {
  leavesOnly: false,
  margin: _Nivo2.default.defaults.margin,
  onBubbleClick: function onBubbleClick() {},
  identity: 'id',
  value: 'value',
  padding: 1,
  colors: _Nivo2.default.defaults.colorRange,
  // placeholders
  namespace: 'html',
  // border
  borderWidth: 0,
  borderColor: 'inherit',
  // labels
  enableLabel: true,
  label: 'id',
  labelTextColor: 'inherit:darker(1)',
  labelTextDY: 5,
  labelSkipRadius: 0,
  // transitions
  animate: true,
  motionStiffness: _Nivo2.default.defaults.motionStiffness, // react-motion
  motionDamping: _Nivo2.default.defaults.motionDamping, // react-motion
  transitionDuration: _Nivo2.default.defaults.transitionDuration, // d3 transitions
  transitionEasing: _Nivo2.default.defaults.transitionEasing };
