// nivo/lib/components/charts/bars/BarItemLegend.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BarItem = function (_Component) {
    _inherits(BarItem, _Component);

    function BarItem() {
        _classCallCheck(this, BarItem);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(BarItem).apply(this, arguments));
    }

    _createClass(BarItem, [{
        key: 'render',
        value: function render() {
            var _props = this.props;
            var x = _props.x;
            var y = _props.y;
            var width = _props.width;
            var height = _props.height;
            var color = _props.color;


            return _react2.default.createElement('g', {
                transform: 'translate(' + x + ',' + y,
                className: 'nivo_bar_legend'
            });
        }
    }]);

    return BarItem;
}(_react.Component);

BarItem.propTypes = {
    value: _react.PropTypes.number.isRequired,
    x: _react.PropTypes.number.isRequired,
    y: _react.PropTypes.number.isRequired,
    width: _react.PropTypes.number.isRequired,
    height: _react.PropTypes.number.isRequired,
    color: _react.PropTypes.string.isRequired
};

BarItem.defaultProps = {};

exports.default = BarItem;
