// nivo/lib/components/charts/bars/Bars.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _ColorUtils = require('../../../ColorUtils');

var _propertiesConverters = require('../../../lib/propertiesConverters');

var _BarItem = require('./BarItem');

var _BarItem2 = _interopRequireDefault(_BarItem);

var _BarItemLabel = require('./BarItemLabel');

var _BarItemLabel2 = _interopRequireDefault(_BarItemLabel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Bars = function (_Component) {
    _inherits(Bars, _Component);

    function Bars() {
        _classCallCheck(this, Bars);

        return _possibleConstructorReturn(this, (Bars.__proto__ || Object.getPrototypeOf(Bars)).apply(this, arguments));
    }

    _createClass(Bars, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                data = _props.data,
                scales = _props.scales,
                _xScale = _props.xScale,
                _yScale = _props.yScale,
                width = _props.width,
                height = _props.height,
                color = _props.color,
                x = _props.x,
                y = _props.y;


            var xScale = scales[_xScale];
            var yScale = scales[_yScale];

            var getX = _lodash2.default.isFunction(x) ? x : function (d) {
                return d[x];
            };
            var getY = _lodash2.default.isFunction(y) ? y : function (d) {
                return d[y];
            };

            var rects = [];
            data.forEach(function (d, i) {
                var x = void 0;
                var y = void 0;
                var barWidth = void 0;
                var barHeight = void 0;

                if (xScale.bandwidth) {
                    barWidth = xScale.bandwidth(); // / series.length
                    x = xScale(getX(d)); // + barWidth * serieIndex
                } else {
                    x = 0;
                    barWidth = xScale(getX(d));
                }

                if (yScale.bandwidth) {
                    barHeight = yScale.bandwidth(); // / series.length
                    y = yScale(getY(d)); // + barHeight * serieIndex
                } else {
                    y = yScale(getY(d));
                    barHeight = height - y;
                }

                if (barWidth > 0 && barHeight > 0) {
                    rects.push({
                        key: 'bar.' + i,
                        //value,
                        x: x, y: y,
                        width: barWidth,
                        height: barHeight,
                        color: color });
                }
            });

            return _react2.default.createElement(
                'g',
                { className: 'bars' },
                rects.map(function (d) {
                    return _react2.default.createElement(_BarItem2.default, _extends({}, d, { key: d.key }));
                })
            );
        }
    }]);

    return Bars;
}(_react.Component);

Bars.propTypes = {
    colors: _react.PropTypes.any.isRequired,
    scales: _react.PropTypes.object,
    width: _react.PropTypes.number,
    height: _react.PropTypes.number,
    // labels
    enableLabels: _react.PropTypes.bool.isRequired,
    // motion
    animate: _react.PropTypes.bool.isRequired,
    motionStiffness: _react.PropTypes.number.isRequired,
    motionDamping: _react.PropTypes.number.isRequired
};
Bars.defaultProps = {
    colors: _Nivo2.default.defaults.colorRange,
    // labels
    enableLabels: true,
    // motion
    animate: true,
    motionStiffness: _Nivo2.default.defaults.motionStiffness,
    motionDamping: _Nivo2.default.defaults.motionDamping
};
exports.default = Bars;
