// nivo/lib/components/Chart.js
/*
 * This file is part of the nivo project.
 *
 * (c) Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDimensions = require('react-dimensions');

var _reactDimensions2 = _interopRequireDefault(_reactDimensions);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Nivo = require('../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Chart = function (_Component) {
    _inherits(Chart, _Component);

    function Chart() {
        _classCallCheck(this, Chart);

        return _possibleConstructorReturn(this, (Chart.__proto__ || Object.getPrototypeOf(Chart)).apply(this, arguments));
    }

    _createClass(Chart, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                containerWidth = _props.containerWidth,
                containerHeight = _props.containerHeight,
                children = _props.children;


            var margin = _lodash2.default.assign({}, _Nivo2.default.defaults.margin, this.props.margin);
            var width = containerWidth - margin.left - margin.right;
            var height = containerHeight - margin.top - margin.bottom;

            return _react2.default.createElement(
                'svg',
                { width: containerWidth, height: containerHeight },
                _react2.default.createElement(
                    'g',
                    { transform: 'translate(' + margin.left + ', ' + margin.top + ')' },
                    _react2.default.Children.map(children, function (child) {
                        return _react2.default.cloneElement(child, { width: width, height: height });
                    })
                )
            );
        }
    }]);

    return Chart;
}(_react.Component);

Chart.displayName = 'Chart';

Chart.propTypes = {
    containerWidth: _react.PropTypes.number.isRequired,
    containerHeight: _react.PropTypes.number.isRequired,
    margin: _react.PropTypes.shape({
        top: _react.PropTypes.number,
        right: _react.PropTypes.number,
        bottom: _react.PropTypes.number,
        left: _react.PropTypes.number
    }).isRequired
};

Chart.defaultProps = {
    margin: _Nivo2.default.defaults.margin
};

exports.default = (0, _reactDimensions2.default)()(Chart);
