// nivo/lib/properties/curve.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.curvePropType = exports.curvePropKeys = exports.curvePropMapping = undefined;

var _react = require('react');

var _d = require('d3');

var curvePropMapping = exports.curvePropMapping = {
    basis: _d.curveBasis,
    basisClosed: _d.curveBasisClosed,
    basisOpen: _d.curveBasisOpen,
    bundle: _d.curveBundle,
    cardinal: _d.curveCardinal,
    cardinalClosed: _d.curveCardinalClosed,
    cardinalOpen: _d.curveCardinalOpen,
    catmullRom: _d.curveCatmullRom,
    catmullRomClosed: _d.curveCatmullRomClosed,
    catmullRomOpen: _d.curveCatmullRomOpen,
    linear: _d.curveLinear,
    linearClosed: _d.curveLinearClosed,
    monotoneX: _d.curveMonotoneX,
    monotoneY: _d.curveMonotoneY,
    natural: _d.curveNatural,
    step: _d.curveStep,
    stepAfter: _d.curveStepAfter,
    stepBefore: _d.curveStepBefore
};

var curvePropKeys = exports.curvePropKeys = Object.keys(curvePropMapping);

var curvePropType = exports.curvePropType = _react.PropTypes.oneOf(curvePropKeys);
