// mozaik-ext-github/lib/components/stats/RepoCommitActivity.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _RepoCommitActivityHistogramChart = require('./charts/RepoCommitActivityHistogramChart');

var _RepoCommitActivityHistogramChart2 = _interopRequireDefault(_RepoCommitActivityHistogramChart);

var _RepoCommitActivityLineChart = require('./charts/RepoCommitActivityLineChart');

var _RepoCommitActivityLineChart2 = _interopRequireDefault(_RepoCommitActivityLineChart);

var _ui = require('mozaik/ui');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RepositoryCommitActivity = function (_Component) {
    _inherits(RepositoryCommitActivity, _Component);

    function RepositoryCommitActivity() {
        _classCallCheck(this, RepositoryCommitActivity);

        return _possibleConstructorReturn(this, (RepositoryCommitActivity.__proto__ || Object.getPrototypeOf(RepositoryCommitActivity)).apply(this, arguments));
    }

    _createClass(RepositoryCommitActivity, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                repository = _props.repository,
                title = _props.title,
                type = _props.type,
                apiData = _props.apiData,
                apiError = _props.apiError;


            var body = _react2.default.createElement(_ui.WidgetLoader, null);
            if (apiData && !apiError && apiData.buckets) {
                if (type === 'histogram') {
                    body = _react2.default.createElement(_RepoCommitActivityHistogramChart2.default, { commits: apiData.buckets });
                } else if (type === 'line') {
                    body = _react2.default.createElement(_RepoCommitActivityLineChart2.default, { commits: apiData.buckets });
                }
            }

            return _react2.default.createElement(
                _ui.Widget,
                null,
                _react2.default.createElement(_ui.WidgetHeader, {
                    title: title || 'Commit Activity',
                    subject: title ? null : repository,
                    icon: 'line-chart'
                }),
                _react2.default.createElement(
                    _ui.WidgetBody,
                    { style: { overflowY: 'hidden' } },
                    _react2.default.createElement(
                        _ui.TrapApiError,
                        { error: apiError },
                        body
                    )
                )
            );
        }
    }], [{
        key: 'getApiRequest',
        value: function getApiRequest(_ref) {
            var repository = _ref.repository;

            return {
                id: 'github.repoCommitActivity.' + repository,
                params: { repository: repository }
            };
        }
    }]);

    return RepositoryCommitActivity;
}(_react.Component);

RepositoryCommitActivity.propTypes = {
    repository: _react.PropTypes.string.isRequired,
    title: _react.PropTypes.string,
//    apiData: _react.PropTypes.shape({
//        buckets: _react.PropTypes.
//    }),
    apiError: _react.PropTypes.object,
    type: _react.PropTypes.oneOf(['histogram', 'line']).isRequired
};
exports.default = RepositoryCommitActivity;
