// nivo/lib/components/charts/pie/Pie.js
/*
 * This file is part of the nivo project.
 *
 * (c) Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _PieD3Svg = require('../../../lib/charts/pie/PieD3Svg');

var _PieD3Svg2 = _interopRequireDefault(_PieD3Svg);

var _decoratorsFromReactChildren = require('../../../lib/decoratorsFromReactChildren');

var _decoratorsFromReactChildren2 = _interopRequireDefault(_decoratorsFromReactChildren);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Pie = function (_Component) {
    _inherits(Pie, _Component);

    function Pie() {
        _classCallCheck(this, Pie);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(Pie).apply(this, arguments));
    }

    _createClass(Pie, [{
        key: 'shouldComponentUpdate',
        value: function shouldComponentUpdate(nextProps) {
            this.pie.decorate((0, _decoratorsFromReactChildren2.default)(nextProps.children, 'decoratePie'));
            this.pie.draw(nextProps);

            return false;
        }
    }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.pie = (0, _PieD3Svg2.default)((0, _reactDom.findDOMNode)(this));
            this.pie.decorate((0, _decoratorsFromReactChildren2.default)(this.props.children, 'decoratePie'));
            this.pie.draw(this.props);
        }
    }, {
        key: 'render',
        value: function render() {
            return _react2.default.createElement('g', null);
        }
    }]);

    return Pie;
}(_react.Component);

var array = _react.PropTypes.array;
var number = _react.PropTypes.number;
var string = _react.PropTypes.string;
var func = _react.PropTypes.func;
var any = _react.PropTypes.any;


Pie.propTypes = {
    width: number.isRequired,
    height: number.isRequired,
    sort: func,
    data: array.isRequired,
    keyProp: string.isRequired,
    valueProp: string.isRequired,
    startAngle: number.isRequired,
    endAngle: number.isRequired,
    padAngle: number.isRequired,
    cornerRadius: number.isRequired,
    colors: any.isRequired,
    transitionDuration: number.isRequired,
    transitionEasing: string.isRequired,
    innerRadius: number.isRequired
};

Pie.defaultProps = {
    sort: null,
    keyProp: 'label',
    valueProp: 'value',
    startAngle: 0,
    endAngle: 360,
    padAngle: 0,
    cornerRadius: 0,
    transitionDuration: _Nivo2.default.defaults.transitionDuration,
    transitionEasing: _Nivo2.default.defaults.transitionEasing,
    innerRadius: 0,
    colors: _Nivo2.default.defaults.colorRange
};

exports.default = Pie;
