// nivo/lib/components/axes/AxisTick.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AxisTick = function (_Component) {
    _inherits(AxisTick, _Component);

    function AxisTick() {
        _classCallCheck(this, AxisTick);

        return _possibleConstructorReturn(this, (AxisTick.__proto__ || Object.getPrototypeOf(AxisTick)).apply(this, arguments));
    }

    _createClass(AxisTick, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                _value = _props.value,
                x = _props.x,
                y = _props.y,
                format = _props.format,
                opacity = _props.opacity,
                tickLine = _props.tickLine,
                textXY = _props.textXY,
                textDY = _props.textDY,
                textAnchor = _props.textAnchor;


            var theme = this.context.nivoTheme;

            var value = _value;
            if (format !== undefined) {
                value = format(value);
            }

            return _react2.default.createElement(
                'g',
                {
                    transform: 'translate(' + x + ',' + y + ')',
                    style: { opacity: opacity }
                },
                _react2.default.createElement('line', _extends({}, tickLine, {
                    stroke: theme.axis.tickColor
                })),
                _react2.default.createElement(
                    'text',
                    _extends({}, textXY, {
                        dy: textDY,
                        textAnchor: textAnchor,
                        style: {
                            fill: theme.axis.textColor,
                            fontSize: theme.axis.fontSize
                        }
                    }),
                    value
                )
            );
        }
    }]);

    return AxisTick;
}(_react.Component);

AxisTick.propTypes = {
    format: _react.PropTypes.func
};
AxisTick.contextTypes = {
    nivoTheme: _react.PropTypes.object.isRequired
};
exports.default = AxisTick;
