// nivo/lib/components/axes/GridLines.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactMotion = require('react-motion');

var _Nivo = require('../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _GridLine = require('./GridLine');

var _GridLine2 = _interopRequireDefault(_GridLine);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var GridLines = function (_Component) {
    _inherits(GridLines, _Component);

    function GridLines(props) {
        _classCallCheck(this, GridLines);

        var _this = _possibleConstructorReturn(this, (GridLines.__proto__ || Object.getPrototypeOf(GridLines)).call(this, props));

        _this.willEnter = _this.willEnter.bind(_this);
        _this.willLeave = _this.willLeave.bind(_this);
        return _this;
    }

    _createClass(GridLines, [{
        key: 'willEnter',
        value: function willEnter(_ref) {
            var style = _ref.style;
            var type = this.props.type;


            return {
                opacity: 0,
                x1: type === 'x' ? 0 : style.x1.val,
                x2: type === 'x' ? 0 : style.x2.val,
                y1: type === 'y' ? 0 : style.y1.val,
                y2: type === 'y' ? 0 : style.y2.val
            };
        }
    }, {
        key: 'willLeave',
        value: function willLeave(_ref2) {
            var style = _ref2.style;
            var _props = this.props,
                motionStiffness = _props.motionStiffness,
                motionDamping = _props.motionDamping;

            var springConfig = {
                stiffness: motionStiffness,
                damping: motionDamping
            };

            return {
                opacity: (0, _reactMotion.spring)(0, springConfig),
                x1: (0, _reactMotion.spring)(style.x1.val, springConfig),
                x2: (0, _reactMotion.spring)(style.x2.val, springConfig),
                y1: (0, _reactMotion.spring)(style.y1.val, springConfig),
                y2: (0, _reactMotion.spring)(style.y2.val, springConfig)
            };
        }
    }, {
        key: 'render',
        value: function render() {
            var _props2 = this.props,
                lines = _props2.lines,
                animate = _props2.animate,
                motionStiffness = _props2.motionStiffness,
                motionDamping = _props2.motionDamping;
            var theme = this.context.nivoTheme;


            if (!animate) {
                return _react2.default.createElement(
                    'g',
                    null,
                    lines.map(function (line) {
                        return _react2.default.createElement(_GridLine2.default, _extends({}, line, theme.grid));
                    })
                );
            }

            var springConfig = {
                stiffness: motionStiffness,
                damping: motionDamping
            };

            return _react2.default.createElement(
                _reactMotion.TransitionMotion,
                {
                    willEnter: this.willEnter,
                    willLeave: this.willLeave,
                    styles: lines.map(function (line) {
                        return {
                            key: line.key,
                            style: {
                                opacity: (0, _reactMotion.spring)(1, springConfig),
                                x1: (0, _reactMotion.spring)(line.x1 || 0, springConfig),
                                x2: (0, _reactMotion.spring)(line.x2 || 0, springConfig),
                                y1: (0, _reactMotion.spring)(line.y1 || 0, springConfig),
                                y2: (0, _reactMotion.spring)(line.y2 || 0, springConfig)
                            }
                        };
                    })
                },
                function (interpolatedStyles) {
                    return _react2.default.createElement(
                        'g',
                        null,
                        interpolatedStyles.map(function (interpolatedStyle) {
                            var key = interpolatedStyle.key,
                                style = interpolatedStyle.style;


                            return _react2.default.createElement(_GridLine2.default, _extends({
                                key: key
                            }, theme.grid, style));
                        })
                    );
                }
            );
        }
    }]);

    return GridLines;
}(_react.Component);

GridLines.propTypes = {
    type: _react.PropTypes.oneOf(['x', 'y']).isRequired,
    lines: _react.PropTypes.arrayOf(_react.PropTypes.shape({
        key: _react.PropTypes.string.isRequired,
        x1: _react.PropTypes.number,
        x2: _react.PropTypes.number,
        y1: _react.PropTypes.number,
        y2: _react.PropTypes.number
    })).isRequired,
    // motion
    animate: _react.PropTypes.bool.isRequired,
    motionStiffness: _react.PropTypes.number.isRequired,
    motionDamping: _react.PropTypes.number.isRequired
};
GridLines.defaultProps = {
    // motion
    animate: true,
    motionStiffness: _Nivo2.default.defaults.motionStiffness,
    motionDamping: _Nivo2.default.defaults.motionDamping
};
GridLines.contextTypes = {
    nivoTheme: _react.PropTypes.object.isRequired
};
exports.default = GridLines;
