// nivo/lib/components/charts/treemap/TreeMapProps.js
/*
 * This file is part of the nivo project.
 *
 * (c) 2016 Raphaël Benitte
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.treeMapDefaultProps = exports.treeMapPropTypes = undefined;

var _react = require('react');

var _Nivo = require('../../../Nivo');

var _Nivo2 = _interopRequireDefault(_Nivo);

var _PropTypes = require('../../../PropTypes');

var _TreeMapD = require('../../../lib/charts/treemap/TreeMapD3');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var object = _react.PropTypes.object,
    number = _react.PropTypes.number,
    string = _react.PropTypes.string,
    bool = _react.PropTypes.bool,
    func = _react.PropTypes.func,
    any = _react.PropTypes.any,
    oneOf = _react.PropTypes.oneOf,
    oneOfType = _react.PropTypes.oneOfType;

/**
 * TreeMap components propTypes.
 *
 * @type {object}
 */

var treeMapPropTypes = exports.treeMapPropTypes = {
  width: number.isRequired, // for non responsive components
  height: number.isRequired, // for non responsive components
  margin: _PropTypes.margin,
  root: object.isRequired, // data
  leavesOnly: bool.isRequired,
  value: oneOfType([string, func]).isRequired,
  identity: oneOfType([string, func]).isRequired,
  tile: oneOf(Object.keys(_TreeMapD.tilingMethods)),

  // labels
  enableLabels: bool.isRequired,
  orientLabels: bool.isRequired,
  label: oneOfType([string, func]).isRequired,
  labelFormat: string,
  labelSkipSize: number.isRequired,

  innerPadding: number.isRequired,
  outerPadding: number.isRequired,
  colors: any.isRequired,

  borderWidth: number.isRequired,
  borderColor: any.isRequired,
  // transitions
  animate: bool.isRequired,
  motionStiffness: number.isRequired, // react-motion
  motionDamping: number.isRequired, // react-motion
  transitionDuration: number.isRequired, // d3 transitions
  transitionEasing: string.isRequired, // d3 transitions

  // placeholders
  namespace: oneOf(['html', 'svg']).isRequired,
  children: func.isRequired
};

/**
 * TreeMap components defaultProps.
 *
 * @type {object}
 */
var treeMapDefaultProps = exports.treeMapDefaultProps = {
  leavesOnly: false,
  tile: 'squarify',
  value: 'value',
  identity: 'name',

  // labels
  enableLabels: true,
  orientLabels: true,
  label: 'id',
  labelSkipSize: 0,
  labelTextColor: 'inherit:darker(1)',

  margin: _Nivo2.default.defaults.margin,
  innerPadding: 0,
  outerPadding: 0,

  borderWidth: 0,
  borderColor: 'inherit',
  // transitions
  animate: true,
  transitionDuration: _Nivo2.default.defaults.transitionDuration,
  transitionEasing: _Nivo2.default.defaults.transitionEasing,
  motionStiffness: 120,
  motionDamping: 10,

  colors: _Nivo2.default.defaults.colorRange
};
